File: codes/python/480.滑动窗口中位数.py


import heapq

class DualHeap:
	def __init__(self, k):
		# self.small是一个大根堆，保存较小的一半元素
		self.small = list()
		# self.large是一个小根堆，保存较大的一半元素
		self.large = list()
		# 哈希表，保存延迟删除的元素，key为元素，value为需要删除的次数
		self.delayed = {}
		self.k = k
		# small和large中元素的个数，需要扣除被延迟删除的元素
		self.small_size = 0
		self.large_size = 0

	def prune(self, heap):
		while heap:
			num = heap[0]
			if heap is self.small:
				num = -num
			if num in self.delayed:
				heapq.heappop(heap)
				self.delayed[num] -= 1
				if self.delayed[num] == 0:
					self.delayed.pop(num)
			else:
				break

	def make_balance(self):
		if self.small_size > self.large_size + 1:
			# small比large多2个元素
			heapq.heappush(self.large, -self.small[0])
			heapq.heappop(self.small)
			self.small_size -= 1
			self.large_size += 1
			# small堆顶元素被移除，需要进行 prune
			self.prune(self.small)
		elif self.small_size < self.large_size:
			# large比small多1个元素
			heapq.heappush(self.small, -self.large[0])
			heapq.heappop(self.large)
			self.small_size += 1
			self.large_size -= 1
			# large堆顶元素被移除，需要进行 prune
			self.prune(self.large)

	def insert(self, num):
		if not self.small or num <= -self.small[0]:
			heapq.heappush(self.small, -num)
			self.small_size += 1
		else:
			heapq.heappush(self.large, num)
			self.large_size += 1
		self.make_balance()

	def erase(self, num):
		self.delayed[num] = self.delayed.get(num, 0) + 1
		if num <= -self.small[0]:
			self.small_size -= 1
			if num == -self.small[0]:
				self.prune(self.small)
		else:
			self.large_size -= 1
			if num == self.large[0]:
				self.prune(self.large)
		self.make_balance()

	def get_median(self):
		return float(-self.small[0]) if self.k & 1 else (-self.small[0] + self.large[0]) / 2
	
class Solution:
	def medianSlidingWindow(self, nums, k):
		dual_heap = DualHeap(k)
		for num in nums[:k]:
			dual_heap.insert(num)
		ans = [dual_heap.get_median()]
		for i in range(k, len(nums)):
			dual_heap.insert(nums[i])
			dual_heap.erase(nums[i - k])
			ans.append(dual_heap.get_median())
		return ans

File: codes/golang/480.滑动窗口中位数.go


import "container/heap"

// 可以设置为一个大根堆或者一个小根堆
type hp struct {
	sort.IntSlice
	comparator func(int, int) bool
}

func (h hp) Less(i, j int) bool { return h.comparator(h.IntSlice[i], h.IntSlice[j]) }

func (h *hp) Push(v interface{}) { h.IntSlice = append(h.IntSlice, v.(int)) }

func (h *hp) Pop() interface{} {
	v := h.IntSlice[h.Len()-1]
	h.IntSlice = h.IntSlice[:h.Len()-1]
	return v
}

type DualHeap struct {
	// 大根堆，维护较小的一半元素
	small *hp
	// 小根堆，维护较大的一半元素
	large *hp
	// 哈希表，记录延迟删除的元素，key为元素，value为需要删除的次数
	delayed map[int]int
	// small 和 large 当前包含的元素个数，需要扣除被删除的元素
	smallSize, largeSize int
	k                    int
}

func NewDualHeap(k int) *DualHeap {
	return &DualHeap{
		small:   &hp{comparator: func(i, j int) bool { return i > j }},
		large:   &hp{comparator: func(i, j int) bool { return i < j }},
		delayed: map[int]int{},
		k:       k,
	}
}

func (h *DualHeap) prune() {
	for h.small.Len() > 0 {
		num := h.small.IntSlice[0]
		if h.delayed[num] > 0 {
			h.delayed[num]--
			if h.delayed[num] == 0 {
				delete(h.delayed, num)
			}
			heap.Pop(h.small)
		} else {
			break
		}
	}
	for h.large.Len() > 0 {
		num := h.large.IntSlice[0]
		if h.delayed[num] > 0 {
			h.delayed[num]--
			if h.delayed[num] == 0 {
				delete(h.delayed, num)
			}
			heap.Pop(h.large)
		} else {
			break
		}
	}
}

func (h *DualHeap) Insert(num int) {
	if h.small.Len() == 0 || num <= h.small.IntSlice[0] {
		heap.Push(h.small, num)
		h.smallSize++
	} else {
		heap.Push(h.large, num)
		h.largeSize++
	}
	h.makeBalance()
}

func (h *DualHeap) Erase(num int) {
	h.delayed[num]++
	if num <= h.small.IntSlice[0] {
		h.smallSize--
		if num == h.small.IntSlice[0] {
			h.prune()
		}
	} else {
		h.largeSize--
		if num == h.large.IntSlice[0] {
			h.prune()
		}
	}
	h.makeBalance()
}

func (h *DualHeap) GetMedian() float64 {
	if h.k&1 == 1 {
		return float64(h.small.IntSlice[0])
	}
	return float64(h.small.IntSlice[0]+h.large.IntSlice[0]) / 2
}

func (h *DualHeap) makeBalance() {
	if h.smallSize > h.largeSize+1 {
		heap.Push(h.large, heap.Pop(h.small))
		h.smallSize--
		h.largeSize++
		h.prune()
	} else if h.smallSize < h.largeSize {
		heap.Push(h.small, heap.Pop(h.large))
		h.smallSize++
		h.largeSize--
		h.prune()
	}
}

func medianSlidingWindow(nums []int, k int) []float64 {
	dh := NewDualHeap(k)
	for _, num := range nums[:k] {
		dh.Insert(num)
	}
	ans := make([]float64, 1, len(nums)-k+1)
	ans[0] = dh.GetMedian()
	for i := k; i < len(nums); i++ {
		dh.Insert(nums[i])
		dh.Erase(nums[i-k])
		ans = append(ans, dh.GetMedian())
	}
	return ans
}
